### PR TITLE
API to resolve redirects for single work

### LINKED ISSUE #5999: Resolve Redirects: Endpoint for updating a single work of /type/redirect
Subtask of #2553

Create an endpoint... say `openlibrary.org/works/merge/finalize.json?records=OL123W,OL145W` (this follows the pattern of the works/merge urls). The endpoint will:

1) Ensure the user is an admin. Limit to only admins for now.
2) Ensure all but one of the provided OLIDs are `/type/redirect`. If not, error.
3) Update reading log table and ratings table (See #5421)
4) Update observations table

That will let us:
1) Test the DB code in isolation ; this is the most dangerous code since any errors here are irreversible!
2) Let us begin writing scripts to fix all the currently broken reading logs; unblocking #2553
3) Let us decouple updating references from work merging. Updating the DB refs effectively makes the merge permanent, because it's irreversible (if we change id A → B and D → B in the database, we can't reverse that operation, because we don't know when we see `B` whether it should be reverted to `A` or `D` or remain `B`!). Having the reference updating be its own endpoint places us in a position to do something like Wikidata, which updates references after a ~1 day window, to allow for undoing.

### Stakeholders

@BharatKalluri

### PR DESCRIPTION
**Pull request by mekarpeles**
- Each feature type (i.e. bookshelves, ratings, observations, booknotes) inherits from db.CommonExtras
- Result shows occurrences of redirect works per type (bookshelves, ratings, observations, booknotes)
- Handles chains of redirects
- _test=True / ?test=true flag to db rollback/abort transaction

Closes #5999 

### Technical

**NB** For any given category (e.g. ratings, booknotes, bookshelves, observations -- let's pick ratings), there is an edge case that a work of type redirect (e.g. work OL123W) will need to resolve to its active work (e.g. OL234W) but this active work already has a db entry in the db (e.g. if the patron has a rating in the db for both OL123W and OL234W). In this case, the transaction of updating the redirect work (e.g. OL123W) to the active work (e.g. OL234W) will fail because of a `UniqueConstraint`. Therefore, in the case where this happens, we must fetch all of the rows where the redirect work (e.g. OL123W) appears in the table (e.g. ratings) and attempt to update them one at a time. If/when the `UniqueConstraint` is encountered, we may then simply delete the row for the redirect work because it has already been accounted for by the existence of the (conflicting) active work.

`db.CommonExtras` parent class was added to core/db so that the behavior for syncing/updating a work_id in any of the tables is the same. Each child class (i.e. ratings, observations, booknotes, bookshelves) now defines two new class constants: `TABLENAME` and `PRIMARY_KEY` (i.e. the key names consisting of the primary key in the db). These two new fields let us make the functions within the `db.CommonExtras` class generic.

### Testing

http://ol-dev1.us.archive.org:1337/admin/resolve_redirects?key=/works/OL5976342W&debug=true&test=true

remove `test=true` to commit the db transaction

### Stakeholders

@cdrini @jimchamp

### CODE REVIEW

File: openlibrary/plugins/admin/code.py [hunk 1]
```diff
@@ -201,6 +205,56 @@ def POST(self):
...
+                        "key": thing.key,
+                        "occurrences": {},
+                        "updates": {},
+                    })
+                    key = thing.location
+                else:
+                    summary['resolved_key'] = thing.key
+
+            for i, _ in enumerate(summary['redirect_chain']):
+                r = summary['redirect_chain'][i]
```
Comment (cdrini) [inline openlibrary/plugins/admin/code.py:231]:
```suggestion
            for r in summary['redirect_chain']:
```

File: openlibrary/plugins/admin/code.py [hunk 2]
```diff
@@ -201,6 +205,56 @@ def POST(self):
...
+                thing = web.ctx.site.get(key)
+                if thing.type.key == "/type/redirect":
+                    summary['redirect_chain'].append({
+                        "key": thing.key,
+                        "occurrences": {},
+                        "updates": {},
+                    })
+                    key = thing.location
+                else:
+                    summary['resolved_key'] = thing.key
```
Comment (cdrini) [inline openlibrary/plugins/admin/code.py:228]: Might be best to pull out into a function like `def find_redirect_chain(key: str) -> list[str]` return just the `keys`

```py
summary['redirect_chain'] = [
    { 'key': k, 'occurrences': {}, 'updates': {} }
    for k in first_redirect_chain(key)
]
```

File: openlibrary/core/db.py [hunk 3]
```diff
@@ -17,6 +17,58 @@ def get_db():
     return _get_db()
 
 
+class CommonExtras:
+    """
+    A set of methods used by bookshelves, booknotes, ratings, and observations tables
+    """
+
+    @classmethod
+    def update_work_id(cls, current_work_id, new_work_id, _test=False):
```
Comment (cdrini) [inline openlibrary/core/db.py:26]: Unit test for this would be killer, since this is irreversible.

**Review by cdrini — COMMENTED** [review]
Some code quality cleanups

It would be nice if in a future PR we made all the e.g. `get_all_works_ratings` methods inherit signature from CommonExtras ; then we can collapse some of the duplication with for loops. But not a blocker.


Need to test this locally to confirm first

**Comment by mekarpeles** [comment]
Test Cases:
1. create 10 records
2. Make sure the right n rows get identified
3. Make sure the right n rows get resolved
4. Make sure the right n rows get deleted
5. **If conflict/collision occurs, make sure the result is what we expect**

**Comment by mekarpeles** [comment]
Test
```
work1_key = "/works/OL123W"
work2_key = "/works/OL234W"
work3_key = "/works/OL345W"
work4_key = "/works/OL456W"

type_redir = {"key": "/type/redirect"}
type_work = {"key": "/type/work"}
make_redir = lambda key1, key2: {"key": key1, "location": key2, "type": type_redir}
work1 = make_redir(work1_key, work2_key)
work2 = make_redir(work2_key, work3_key)
work3 = make_redir(work3_key, work4_key)
work4 = {"key": work4_key, "type": type_work}

web.ctx = web.storage()
web.ctx.site = {
    work1_key: work1,
    work2_key: work2,
    work3_key: work3,
    work4_key: work4
}

class TestResolveRedirectsChain:
    def setup_method(self):
        pass

    def test_resolve_redirect_chain(self):
        work_key = "/works/OL123W"
        redirect_chain = resolve_redirects(work_key)
        assert redirect_chain
        resolved_work = redirect_chain[-1]
        assert resolved_work.get("type") == type_work
        assert resolved_work.get("key") == work4_key


def resolve_redirects(work_key):
    resolved_key = None
    redirect_chain = []
    key = work_key
    while not resolved_key:
        thing = web.ctx.site.get(key)
        redirect_chain.append(thing)
        if thing['type']['key'] == "/type/redirect":
            key = thing['location']
        else:
            resolved_key = thing['key']
    return redirect_chain
```

File: openlibrary/core/db.py [hunk 4]
```diff
@@ -17,6 +17,58 @@ def get_db():
...
+                    cls.TABLENAME,
+                    where="work_id=$work_id",
+                    vars={"work_id": current_work_id})
+                for row in rows:
+                    where = " AND ".join([
+                        f"{k}='{v}'"
+                        for k, v in row.items()
+                        if k in cls.PRIMARY_KEY
+                    ])
+                    web.debug(where)
```
Comment (mekarpeles) [inline openlibrary/core/db.py:53]:
```suggestion
```

File: openlibrary/core/db.py [hunk 5]
```diff
@@ -17,6 +17,58 @@ def get_db():
...
+                        # otherwise, delete row with current_work_id if failed
+                        oldb.query(f"DELETE FROM {cls.TABLENAME} WHERE {where}")
+                        rows_deleted += 1
+            except:
+                t.rollback()
+                raise
+        except:
+            t.rollback()
+            raise
+        t.rollback() if _test else t.commit()
```
Comment (mekarpeles) [inline openlibrary/core/db.py:68]:
```suggestion
        if _test:
            t.rollback()
        else:
            t.commit()
```

File: openlibrary/plugins/admin/code.py [hunk 6]
```diff
@@ -201,6 +206,53 @@ def POST(self):
...
+                {
+                    "key": thing.key,
+                    "occurrences": {},
+                    "updates": {}
+                } for thing in redirect_chain
+            ]
+            summary['resolved_key'] = redirect_chain[-1].key
+
+            for i, _ in enumerate(summary['redirect_chain']):
+                r = summary['redirect_chain'][i]
```
Comment (mekarpeles) [inline openlibrary/plugins/admin/code.py:229]:
```suggestion
            for r in summary['redirect_chain']:
```

**Comment by cdrini** [comment]
To tests:

- test on a book not on any reading logs
- Test book + redirectOfBook on same reading log shelf
- Test book + redirectOfBook on different reading log shelves

- Test book with ratings
    - Avg rating should change
- Test book with observations

**Review by cdrini — APPROVED** [review]
To tests:

- ✅ test on a book not on any reading logs
- ✅ Test book + redirectOfBook on same reading log shelf
     - one removed, other stays put
- ✅* Test book + redirectOfBook on different reading log shelves
    - Same book now both on Currently Reading and Already read..... I think this is not supposed to happen, but its fine

- ✅ Test book with ratings
    - Avg rating should change
- ✅ Test book with observations
- ✅* Test book with notes
    - We might want to fail here if we have a conflict on UPDATE ; I don't imagine it'll happen often, but deleting someones notes would kinda suck.

**Comment by cdrini** [comment]
Alright lgtm! Take a look at those commits I added just so we're on the same, but otherwise merge at your discretion.
